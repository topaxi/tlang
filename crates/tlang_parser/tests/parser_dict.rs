mod common;

#[test]
fn test_dictionary_literal() {
    assert_parser_snapshot!(r"let x = { foo: 1, bar: 2 };");
}

#[test]
fn test_dictionary_literal_trailing_comma() {
    assert_parser_snapshot!(r"let x = { foo: 1, bar: 2, };");
}

#[test]
fn test_dictionary_literal_shorthand() {
    assert_parser_snapshot!(r"let x = { foo, bar };");
}

#[test]
fn test_dictionary_literal_shorthand_trailing_comma() {
    assert_parser_snapshot!(r"let x = { foo, bar, };");
}

#[test]
fn test_function_call_with_dictionary_no_parens() {
    // TODO: Alternatively use ruby like syntax for calling functions with dictionaries
    //       and omit the braces instead of the parens.
    //       As this syntax here might be more desirable for creating dictionaries or structs.
    assert_parser_snapshot!(r"foo { foo: 1, bar: 2 };");
}

#[test]
#[ignore] // TODO: Decide if we want to use this instead of the above syntax.
fn test_fn_call_expression_with_dictionary_argument() {
    // Similar to ruby syntax, we might want to allow omitting curly braces
    // for the last argument if it is a dictionary (or struct).
    // Creation of the struct/dictionary could/should be infered from the function signature.
    assert_parser_snapshot!("bar(x: 1, y: 2)");
}
