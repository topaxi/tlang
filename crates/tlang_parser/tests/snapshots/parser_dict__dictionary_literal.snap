---
source: crates/tlang_parser/tests/parser_dict.rs
expression: "parse! (r\"let x = { foo: 1, bar: 2 };\")"
---
Module(
  id: NodeId(1),
  statements: [
    Stmt(
      id: NodeId(8),
      kind: Let(LetDeclaration(
        pattern: Pat(
          id: NodeId(2),
          kind: Identifier(Ident(
            name: "x",
            span: Span(
              start: LineColumn(
                index: 4,
                line: 0,
                column: 4,
              ),
              end: LineColumn(
                index: 5,
                line: 0,
                column: 5,
              ),
            ),
          )),
          leading_comments: [],
          trailing_comments: [],
          span: Span(
            start: LineColumn(
              index: 4,
              line: 0,
              column: 4,
            ),
            end: LineColumn(
              index: 5,
              line: 0,
              column: 5,
            ),
          ),
        ),
        expression: Expr(
          id: NodeId(7),
          kind: Dict([
            (Expr(
              id: NodeId(3),
              kind: Path(Path(
                segments: [
                  Ident(
                    name: "foo",
                    span: Span(
                      start: LineColumn(
                        index: 10,
                        line: 0,
                        column: 10,
                      ),
                      end: LineColumn(
                        index: 13,
                        line: 0,
                        column: 13,
                      ),
                    ),
                  ),
                ],
                span: Span(
                  start: LineColumn(
                    index: 10,
                    line: 0,
                    column: 10,
                  ),
                  end: LineColumn(
                    index: 13,
                    line: 0,
                    column: 13,
                  ),
                ),
              )),
              leading_comments: [],
              trailing_comments: [],
              span: Span(
                start: LineColumn(
                  index: 10,
                  line: 0,
                  column: 10,
                ),
                end: LineColumn(
                  index: 13,
                  line: 0,
                  column: 13,
                ),
              ),
            ), Expr(
              id: NodeId(4),
              kind: Literal(UnsignedInteger(1)),
              leading_comments: [],
              trailing_comments: [],
              span: Span(
                start: LineColumn(
                  index: 15,
                  line: 0,
                  column: 15,
                ),
                end: LineColumn(
                  index: 16,
                  line: 0,
                  column: 16,
                ),
              ),
            )),
            (Expr(
              id: NodeId(5),
              kind: Path(Path(
                segments: [
                  Ident(
                    name: "bar",
                    span: Span(
                      start: LineColumn(
                        index: 18,
                        line: 0,
                        column: 18,
                      ),
                      end: LineColumn(
                        index: 21,
                        line: 0,
                        column: 21,
                      ),
                    ),
                  ),
                ],
                span: Span(
                  start: LineColumn(
                    index: 18,
                    line: 0,
                    column: 18,
                  ),
                  end: LineColumn(
                    index: 21,
                    line: 0,
                    column: 21,
                  ),
                ),
              )),
              leading_comments: [],
              trailing_comments: [],
              span: Span(
                start: LineColumn(
                  index: 18,
                  line: 0,
                  column: 18,
                ),
                end: LineColumn(
                  index: 21,
                  line: 0,
                  column: 21,
                ),
              ),
            ), Expr(
              id: NodeId(6),
              kind: Literal(UnsignedInteger(2)),
              leading_comments: [],
              trailing_comments: [],
              span: Span(
                start: LineColumn(
                  index: 23,
                  line: 0,
                  column: 23,
                ),
                end: LineColumn(
                  index: 24,
                  line: 0,
                  column: 24,
                ),
              ),
            )),
          ]),
          leading_comments: [],
          trailing_comments: [],
          span: Span(
            start: LineColumn(
              index: 8,
              line: 0,
              column: 8,
            ),
            end: LineColumn(
              index: 26,
              line: 0,
              column: 26,
            ),
          ),
        ),
        type_annotation: None,
      )),
      span: Span(
        start: LineColumn(
          index: 0,
          line: 0,
          column: 0,
        ),
        end: LineColumn(
          index: 27,
          line: 0,
          column: 27,
        ),
      ),
      leading_comments: [],
      trailing_comments: [],
    ),
  ],
  span: Span(
    start: LineColumn(
      index: 0,
      line: 0,
      column: 0,
    ),
    end: LineColumn(
      index: 0,
      line: 0,
      column: 0,
    ),
  ),
)
